File: cli/src/schema_compat.rs
//! Target-chart schema compatibility planner for `curie cluster upgrade` (#2588).
//!
//! Ports `curie_api.schema_compat.plan_upgrade` / `render_decision`. The live
//! path reads target metadata from `helm template --show-only
//! templates/schema-compat.yaml`, not from a checkout `include_str!`.

use std::collections::{BTreeMap, BTreeSet};

use serde::Deserialize;

use crate::schema_window::redact_probe_text;

pub const KIND_EXPAND: &str = "expand";
pub const KIND_CONTRACT: &str = "contract";
pub const KIND_IRREVERSIBLE: &str = "irreversible";

#[derive(Debug, Clone, Deserialize)]
pub struct RevisionNode {
    pub revision: String,
    #[serde(default)]
    pub parents: Vec<String>,
    pub kind: String,
    #[serde(default)]
    #[allow(dead_code)]
    pub sha256: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct TargetMetadata {
    pub schema_min: String,
    pub schema_head: String,
    #[serde(default)]
    pub revisions: Vec<RevisionNode>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PendingStep {
    pub revision: String,
    pub kind: String,
}

#[derive(Debug, Clone)]
pub struct CompatDecision {
    pub action: String,
    pub current_revision: Option<String>,
    pub target_head: String,
    pub target_min: String,
    pub pending: Vec<PendingStep>,
    pub rollback_compatible: bool,
    pub reason: String,
    pub forward_only: bool,
    pub outcome: Option<String>,
    pub source_head: Option<String>,
}

/// Parse a helm-template ConfigMap (YAML or JSON) into target metadata.
pub fn parse_target_metadata(rendered: &str) -> Result<TargetMetadata, String> {
    let trimmed = rendered.trim();
    if trimmed.is_empty() {
        return Err(
            "target chart did not render schema compatibility metadata (BYO api.deploy=false or missing template)"
                .into(),
        );
    }
    let mut last_err = None;
    for doc in trimmed.split("\n---") {
        let doc = doc.trim().trim_start_matches("---").trim();
        if doc.is_empty() {
            continue;
        }
        match serde_norway::from_str::<serde_json::Value>(doc) {
            Ok(value) => {
                if let Some(result) = metadata_from_value(&value) {
                    return result;
                }
            }
            Err(error) => last_err = Some(error.to_string()),
        }
    }
    Err(last_err.unwrap_or_else(|| {
        "target chart did not render schema compatibility metadata (BYO api.deploy=false or missing template)"
            .into()
    }))
}

fn metadata_from_value(value: &serde_json::Value) -> Option<Result<TargetMetadata, String>> {
    if let Some(payload) = value
        .get("data")
        .and_then(|data| data.get("compatibility.json"))
    {
        return Some(parse_compatibility_payload(payload));
    }
    if value.get("schema_min").is_some() && value.get("schema_head").is_some() {
        return Some(parse_metadata_object(value));
    }
    None
}

fn parse_compatibility_payload(payload: &serde_json::Value) -> Result<TargetMetadata, String> {
    match payload {
        serde_json::Value::String(raw) => {
            let raw = raw.trim();
            let parsed: serde_json::Value = match serde_json::from_str(raw) {
                Ok(value) => value,
                Err(_) => serde_norway::from_str(raw)
                    .map_err(|e| format!("compatibility.json is not valid JSON: {e}"))?,
            };
            parse_metadata_object(&parsed)
        }
        serde_json::Value::Object(_) => parse_metadata_object(payload),
        _ => Err("compatibility.json is not an object".into()),
    }
}

fn parse_metadata_object(value: &serde_json::Value) -> Result<TargetMetadata, String> {
    serde_json::from_value(value.clone())
        .map_err(|e| format!("schema compatibility metadata is invalid: {e}"))
}

/// Walk pending revisions from `live` (or empty-DB base) to `schema_head`.
///
/// Live walk follows the unique child (parent reverse). Live not in the graph
/// is an incompatible refuse whose reason names the revision.
pub fn pending_revisions(
    live: Option<&str>,
    target: &TargetMetadata,
) -> Result<Vec<PendingStep>, String> {
    match live {
        Some(current) => pending_from_live(current, target),
        None => Ok(pending_from_empty(target)),
    }
}

fn pending_from_empty(target: &TargetMetadata) -> Vec<PendingStep> {
    let by_id: BTreeMap<&str, &RevisionNode> = target
        .revisions
        .iter()
        .map(|node| (node.revision.as_str(), node))
        .collect();
    let mut chain = Vec::new();
    let mut current = Some(target.schema_head.as_str());
    let mut seen = BTreeSet::new();
    while let Some(rev) = current {
        if !seen.insert(rev) {
            break;
        }
        let Some(node) = by_id.get(rev) else {
            break;
        };
        chain.push(PendingStep {
            revision: node.revision.clone(),
            kind: node.kind.clone(),
        });
        current = node
            .parents
            .first()
            .map(String::as_str)
            .filter(|parent| by_id.contains_key(parent));
    }
    chain.reverse();
    chain
}

fn pending_from_live(live: &str, target: &TargetMetadata) -> Result<Vec<PendingStep>, String> {
    let by_id: BTreeMap<&str, &RevisionNode> = target
        .revisions
        .iter()
        .map(|node| (node.revision.as_str(), node))
        .collect();
    if !by_id.contains_key(live) {
        return Err(format!(
            "live database revision {live} is not in the target application's schema graph"
        ));
    }
    if live == target.schema_head {
        return Ok(Vec::new());
    }
    let mut children: BTreeMap<&str, Vec<&RevisionNode>> = BTreeMap::new();
    for node in &target.revisions {
        for parent in &node.parents {
            children.entry(parent.as_str()).or_default().push(node);
        }
    }
    let mut pending = Vec::new();
    let mut current = live;
    let mut seen = BTreeSet::new();
    while current != target.schema_head {
        if !seen.insert(current) {
            return Err(format!("schema graph cycle at revision {current}"));
        }
        let Some(nexts) = children.get(current) else {
            return Err(format!(
                "live database revision {live} cannot reach target head {}",
                target.schema_head
            ));
        };
        if nexts.len() != 1 {
            return Err(format!(
                "schema graph has {} children of revision {current}; expected a unique child",
                nexts.len()
            ));
        }
        let next = nexts[0];
        pending.push(PendingStep {
            revision: next.revision.clone(),
            kind: next.kind.clone(),
        });
        current = next.revision.as_str();
    }
    Ok(pending)
}

/// Pure planner: no database mutation. Matches Python `plan_upgrade`.
pub fn plan_upgrade(
    current_revision: Option<&str>,
    target: &TargetMetadata,
    pending: &[PendingStep],
    forward_only: bool,
    source_head: Option<&str>,
) -> CompatDecision {
    let source = source_head
        .map(ToOwned::to_owned)
        .or_else(|| current_revision.map(ToOwned::to_owned));
    if current_revision.is_none() {
        return CompatDecision {
            action: "apply".into(),
            current_revision: None,
            target_head: target.schema_head.clone(),
            target_min: target.schema_min.clone(),
            pending: pending.to_vec(),
            rollback_compatible: false,
            reason: "empty database; apply migrations to target head".into(),
            forward_only,
            outcome: None,
            source_head: source,
        };
    }
    if current_revision == Some(target.schema_head.as_str()) || pending.is_empty() {
        return CompatDecision {
            action: "noop".into(),
            current_revision: current_revision.map(ToOwned::to_owned),
            target_head: target.schema_head.clone(),
            target_min: target.schema_min.clone(),
            pending: Vec::new(),
            rollback_compatible: true,
            reason: "database is already at the target head".into(),
            forward_only,
            outcome: Some("already_at_head".into()),
            source_head: source,
        };
    }
    let blocking: Vec<&PendingStep> = pending
        .iter()
        .filter(|step| step.kind == KIND_CONTRACT || step.kind == KIND_IRREVERSIBLE)
        .collect();
    if !blocking.is_empty() && !forward_only {
        let names = blocking
            .iter()
            .map(|step| step.revision.as_str())
            .collect::<Vec<_>>()
            .join(", ");
        return CompatDecision {
            action: "refuse".into(),
            current_revision: current_revision.map(ToOwned::to_owned),
            target_head: target.schema_head.clone(),
            target_min: target.schema_min.clone(),
            pending: pending.to_vec(),
            rollback_compatible: false,
            reason: format!(
                "pending contract/irreversible migration {names} would close the \
                 patch rollback window; pass --forward-only / api.migrate.forwardOnly \
                 to apply the documented forward-only procedure"
            ),
            forward_only: false,
            outcome: Some("refused".into()),
            source_head: source,
        };
    }
    let mut rollback_ok = pending.iter().all(|step| step.kind == KIND_EXPAND);
    if forward_only && !blocking.is_empty() {
        rollback_ok = false;
    }
    CompatDecision {
        action: "apply".into(),
        current_revision: current_revision.map(ToOwned::to_owned),
        target_head: target.schema_head.clone(),
        target_min: target.schema_min.clone(),
        pending: pending.to_vec(),
        rollback_compatible: rollback_ok,
        reason: if rollback_ok {
            "pending migrations are expand-only; source application can keep serving".into()
        } else {
            "forward-only apply of a contract/irreversible migration".into()
        },
        forward_only,
        outcome: None,
        source_head: source,
    }
}

pub fn refuse_with_reason(
    current_revision: Option<&str>,
    target: Option<&TargetMetadata>,
    pending: Vec<PendingStep>,
    reason: impl Into<String>,
    forward_only: bool,
    source_head: Option<&str>,
) -> CompatDecision {
    CompatDecision {
        action: "refuse".into(),
        current_revision: current_revision.map(ToOwned::to_owned),
        target_head: target.map(|t| t.schema_head.clone()).unwrap_or_default(),
        target_min: target.map(|t| t.schema_min.clone()).unwrap_or_default(),
        pending,
        rollback_compatible: false,
        reason: reason.into(),
        forward_only,
        outcome: Some("refused".into()),
        source_head: source_head.map(ToOwned::to_owned),
    }
}

/// Structured, redacted planner record. No URLs, passwords, or rows.
pub fn render_decision(decision: &CompatDecision) -> serde_json::Value {
    serde_json::json!({
        "decision": decision.action,
        "current_revision": decision.current_revision,
        "target_min": decision.target_min,
        "target_head": decision.target_head,
        "source_head": decision.source_head,
        "pending": decision.pending.iter().map(|step| {
            serde_json::json!({
                "revision": step.revision,
                "kind": step.kind,
            })
        }).collect::<Vec<_>>(),
        "rollback_compatible": decision.rollback_compatible,
        "forward_only": decision.forward_only,
        "reason": redact_probe_text(&decision.reason),
        "outcome": decision.outcome,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeMap;
    use std::fs;
    use std::path::Path;

    fn kinds() -> BTreeMap<String, String> {
        let path = Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .expect("cli is not the repo root")
            .join("apps/api/src/curie_api/revision_kinds.json");
        serde_json::from_str(&fs::read_to_string(&path).expect("revision_kinds.json"))
            .expect("revision_kinds.json is valid")
    }

    fn kind(rev: &str) -> String {
        kinds()
            .get(rev)
            .cloned()
            .unwrap_or_else(|| KIND_EXPAND.to_string())
    }

    fn node(revision: &str, parent: Option<&str>) -> RevisionNode {
        RevisionNode {
            revision: revision.into(),
            parents: parent.map(|p| vec![p.into()]).unwrap_or_default(),
            kind: kind(revision),
            sha256: None,
        }
    }

    fn contract_target() -> TargetMetadata {
        TargetMetadata {
            schema_min: "0043".into(),
            schema_head: "0043".into(),
            revisions: vec![
                node("0039", Some("0038")),
                node("0040", Some("0039")),
                node("0041", Some("0040")),
                node("0042", Some("0041")),
                node("0043", Some("0042")),
            ],
        }
    }

    fn irreversible_target() -> TargetMetadata {
        TargetMetadata {
            schema_min: "0017".into(),
            schema_head: "0017".into(),
            revisions: vec![
                node("0015", None),
                node("0016", Some("0015")),
                node("0017", Some("0016")),
            ],
        }
    }

    #[test]
    fn kinds_file_grounds_contract_and_irreversible() {
        let kinds = kinds();
        assert_eq!(kinds.get("0041").map(String::as_str), Some(KIND_CONTRACT));
        assert_eq!(
            kinds.get("0016").map(String::as_str),
            Some(KIND_IRREVERSIBLE)
        );
    }

    #[test]
    fn contract_refuses_without_forward_only() {
        let target = contract_target();
        let pending = pending_revisions(Some("0039"), &target).expect("0039 is in graph");
        assert!(
            pending
                .iter()
                .any(|s| s.revision == "0041" && s.kind == KIND_CONTRACT),
            "{pending:?}"
        );
        let decision = plan_upgrade(Some("0039"), &target, &pending, false, Some("0039"));
        assert_eq!(decision.action, "refuse");
        assert!(
            decision.reason.contains("--forward-only"),
            "{}",
            decision.reason
        );
        assert!(decision.reason.contains("0041"), "{}", decision.reason);
        let rendered = render_decision(&decision);
        assert_eq!(rendered["decision"], "refuse");
        assert!(rendered["reason"]
            .as_str()
            .unwrap()
            .contains("--forward-only"));
    }

    #[test]
    fn forward_only_applies_pending_contract() {
        let target = contract_target();
        let pending = pending_revisions(Some("0039"), &target).expect("0039 is in graph");
        let decision = plan_upgrade(Some("0039"), &target, &pending, true, Some("0039"));
        assert_eq!(decision.action, "apply");
        assert!(decision.forward_only);
        assert!(!decision.rollback_compatible);
    }

    #[test]
    fn irreversible_refuses_without_forward_only() {
        let target = irreversible_target();
        let pending = pending_revisions(Some("0015"), &target).expect("0015 is in graph");
        assert!(
            pending
                .iter()
                .any(|s| s.revision == "0016" && s.kind == KIND_IRREVERSIBLE),
            "{pending:?}"
        );
        let decision = plan_upgrade(Some("0015"), &target, &pending, false, None);
        assert_eq!(decision.action, "refuse");
        assert!(decision.reason.contains("--forward-only"));
        assert!(decision.reason.contains("0016"));
    }

    #[test]
    fn empty_database_applies_including_historical_irreversible() {
        let target = irreversible_target();
        let pending = pending_revisions(None, &target).expect("empty walk");
        assert!(pending.iter().any(|s| s.revision == "0016"));
        let decision = plan_upgrade(None, &target, &pending, false, None);
        assert_eq!(decision.action, "apply");
        assert!(!decision.rollback_compatible);
    }

    #[test]
    fn already_at_head_is_noop() {
        let target = contract_target();
        let pending = pending_revisions(Some("0043"), &target).expect("head is in graph");
        assert!(pending.is_empty());
        let decision = plan_upgrade(Some("0043"), &target, &pending, false, Some("0043"));
        assert_eq!(decision.action, "noop");
        assert_eq!(decision.outcome.as_deref(), Some("already_at_head"));
    }

    #[test]
    fn unknown_live_revision_refuses_and_names_it() {
        let target = contract_target();
        let err = pending_revisions(Some("0099"), &target).expect_err("unknown live");
        assert!(err.contains("0099"), "{err}");
        let decision = refuse_with_reason(
            Some("0099"),
            Some(&target),
            Vec::new(),
            err,
            false,
            Some("0099"),
        );
        assert_eq!(decision.action, "refuse");
        assert!(decision.reason.contains("0099"));
    }

    #[test]
    fn parse_configmap_string_or_object_payload() {
        let object = serde_json::json!({
            "kind": "ConfigMap",
            "data": {
                "compatibility.json": {
                    "schema_min": "0043",
                    "schema_head": "0043",
                    "revisions": [{"revision": "0043", "parents": ["0042"], "kind": "expand"}]
                }
            }
        });
        let parsed = parse_target_metadata(&object.to_string()).expect("object payload");
        assert_eq!(parsed.schema_head, "0043");

        let as_string = serde_json::json!({
            "kind": "ConfigMap",
            "data": {
                "compatibility.json": serde_json::json!({
                    "schema_min": "0043",
                    "schema_head": "0043",
                    "revisions": []
                }).to_string()
            }
        });
        let parsed = parse_target_metadata(&as_string.to_string()).expect("string payload");
        assert_eq!(parsed.schema_min, "0043");
    }

    #[test]
    fn render_decision_redacts_probe_text() {
        let decision = refuse_with_reason(
            Some("0039"),
            None,
            Vec::new(),
            "could not connect to postgresql://curie:secret-password@postgres:5432/curie",
            false,
            None,
        );
        let rendered = render_decision(&decision);
        let reason = rendered["reason"].as_str().unwrap();
        assert!(!reason.contains("secret-password"));
        assert!(!reason.contains("postgresql://"));
    }
}

File: apps/api/src/curie_api/schema_compat.py
"""Application/database schema compatibility window (#2300).

A released API image declares the schema range it understands. Migrations run
in one upgrade phase (compose ``curie-migrate``, the chart pre-upgrade Job),
never in every API pod. Patch expands stay rollback-compatible: application
N-1 can serve a newer expand it does not itself migrate.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import logging
import os
import sys
import time
from collections.abc import Iterable, Sequence
from dataclasses import dataclass
from importlib.resources import files
from pathlib import Path
from typing import Any

from alembic import command
from alembic.config import Config
from alembic.script import ScriptDirectory
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine

from .config import get_settings
from .db import SCHEMA

logger = logging.getLogger("curie_api.schema_compat")

KIND_EXPAND = "expand"
KIND_CONTRACT = "contract"
KIND_IRREVERSIBLE = "irreversible"
_VALID_KINDS = {KIND_EXPAND, KIND_CONTRACT, KIND_IRREVERSIBLE}

_WINDOW_RESOURCE = "schema_compat.json"
_KINDS_RESOURCE = "revision_kinds.json"

def _default_alembic() -> Path:
    """Prefer the image copy, then the source tree next to this package."""
    candidates = (
        Path("/app/alembic"),
        Path(__file__).resolve().parents[2] / "alembic",
    )
    for candidate in candidates:
        if candidate.is_dir():
            return candidate
    return Path("/app/alembic")


_DEFAULT_ALEMBIC = _default_alembic()


@dataclass(frozen=True)
class AppWindow:
    schema_min: str
    schema_head: str

    def __post_init__(self) -> None:
        if not self.schema_min or not self.schema_head:
            raise ValueError("schema_min and schema_head must be non-empty")


@dataclass(frozen=True)
class PendingStep:
    revision: str
    kind: str


@dataclass
class CompatDecision:
    action: str
    current_revision: str | None
    target_head: str
    target_min: str
    pending: list[PendingStep]
    rollback_compatible: bool
    reason: str
    forward_only: bool
    outcome: str | None = None
    source_head: str | None = None


def load_window() -> AppWindow:
    payload = json.loads(files("curie_api").joinpath(_WINDOW_RESOURCE).read_text())
    return AppWindow(
        schema_min=str(payload["schema_min"]),
        schema_head=str(payload["schema_head"]),
    )


def load_kinds() -> dict[str, str]:
    payload = json.loads(files("curie_api").joinpath(_KINDS_RESOURCE).read_text())
    kinds = {str(k): str(v) for k, v in payload.items()}
    unknown = sorted({kind for kind in kinds.values() if kind not in _VALID_KINDS})
    if unknown:
        raise ValueError(f"revision_kinds.json has unknown kinds: {unknown}")
    return kinds


def _alembic_config(override: Config | None = None) -> Config:
    if override is not None:
        return override
    cfg = Config()
    cfg.set_main_option("script_location", str(_default_alembic()))
    return cfg


def _script(cfg: Config) -> ScriptDirectory:
    return ScriptDirectory.from_config(cfg)


async def current_revision_async() -> str | None:
    engine = create_async_engine(get_settings().database_url)
    try:
        async with engine.connect() as conn:
            exists = await conn.execute(
                text("SELECT to_regclass(:reg)"),
                {"reg": f"{SCHEMA}.alembic_version"},
            )
            if exists.scalar() is None:
                return None
            rows = await conn.execute(
                text(f"SELECT version_num FROM {SCHEMA}.alembic_version")
            )
            values = [row[0] for row in rows.fetchall()]
    finally:
        await engine.dispose()
    if not values:
        return None
    return str(values[0])


def current_revision() -> str | None:
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.run(current_revision_async())
    raise RuntimeError("current_revision() cannot run inside an event loop")


def _walk_down(script: ScriptDirectory, revision: str) -> list[str]:
    chain: list[str] = []
    seen: set[str] = set()
    rev: str | None = revision
    while rev and rev not in seen:
        seen.add(rev)
        chain.append(rev)
        try:
            rec = script.get_revision(rev)
        except Exception:
            break
        down = rec.down_revision
        next_rev: str | None
        if isinstance(down, (list, tuple)):
            next_rev = str(down[0]) if down else None
        elif down is None:
            next_rev = None
        else:
            next_rev = str(down)
        rev = next_rev
    return chain


def _is_at_or_after(script: ScriptDirectory, current: str, minimum: str) -> bool:
    return minimum in _walk_down(script, current)


def can_serve(
    current: str | None,
    window: AppWindow,
    known_revisions: Iterable[str],
    script: ScriptDirectory | None = None,
) -> bool:
    """True when this application can start against ``current``.

    Unknown future revisions (not in this build's script) are treated as a
    compatible expand so N-1 can keep serving after N's schema expansion.
    A known revision older than ``schema_min`` is not servable.
    """
    if current is None:
        return False
    known = set(known_revisions)
    if current not in known:
        return True
    if current == window.schema_min or current == window.schema_head:
        return True
    if script is None:
        try:
            script = _script(_alembic_config())
        except Exception:
            return False
    return _is_at_or_after(script, current, window.schema_min)


def plan_upgrade(
    *,
    current_revision: str | None,
    window: AppWindow,
    kinds: dict[str, str],
    pending: Sequence[str],
    forward_only: bool,
    source_head: str | None = None,
) -> CompatDecision:
    """Pure planner: no database mutation.

    An empty database (install) applies history, including historical
    irreversible revisions; there is no serving application to protect.
    A live database refuses pending contract/irreversible revisions unless
    ``forward_only`` is set.
    """
    pending_steps = [
        PendingStep(revision=rev, kind=kinds.get(rev, KIND_EXPAND)) for rev in pending
    ]
    source = source_head or current_revision
    if current_revision is None:
        return CompatDecision(
            action="apply",
            current_revision=None,
            target_head=window.schema_head,
            target_min=window.schema_min,
            pending=pending_steps,
            rollback_compatible=False,
            reason="empty database; apply migrations to target head",
            forward_only=forward_only,
            source_head=source,
        )
    if current_revision == window.schema_head or not pending_steps:
        return CompatDecision(
            action="noop",
            current_revision=current_revision,
            target_head=window.schema_head,
            target_min=window.schema_min,
            pending=[],
            rollback_compatible=True,
            reason="database is already at the target head",
            forward_only=forward_only,
            outcome="already_at_head",
            source_head=source,
        )
    blocking = [
        step
        for step in pending_steps
        if step.kind in {KIND_CONTRACT, KIND_IRREVERSIBLE}
    ]
    if blocking and not forward_only:
        names = ", ".join(step.revision for step in blocking)
        return CompatDecision(
            action="refuse",
            current_revision=current_revision,
            target_head=window.schema_head,
            target_min=window.schema_min,
            pending=pending_steps,
            rollback_compatible=False,
            reason=(
                f"pending contract/irreversible migration {names} would close the "
                "patch rollback window; pass --forward-only / api.migrate.forwardOnly "
                "to apply the documented forward-only procedure"
            ),
            forward_only=False,
            outcome="refused",
            source_head=source,
        )
    rollback_ok = all(step.kind == KIND_EXPAND for step in pending_steps)
    if forward_only and blocking:
        rollback_ok = False
    return CompatDecision(
        action="apply",
        current_revision=current_revision,
        target_head=window.schema_head,
        target_min=window.schema_min,
        pending=pending_steps,
        rollback_compatible=rollback_ok,
        reason=(
            "pending migrations are expand-only; source application can keep serving"
            if rollback_ok
            else "forward-only apply of a contract/irreversible migration"
        ),
        forward_only=forward_only,
        source_head=source,
    )


def render_decision(decision: CompatDecision) -> dict[str, Any]:
    """Structured, redacted planner/apply record. No URLs, passwords, or rows."""
    return {
        "decision": decision.action,
        "current_revision": decision.current_revision,
        "target_min": decision.target_min,
        "target_head": decision.target_head,
        "source_head": decision.source_head,
        "pending": [
            {"revision": step.revision, "kind": step.kind} for step in decision.pending
        ],
        "rollback_compatible": decision.rollback_compatible,
        "forward_only": decision.forward_only,
        "reason": decision.reason,
        "outcome": decision.outcome,
    }


def _pending_from_script(
    script: ScriptDirectory, current: str | None, target_head: str
) -> tuple[str, ...]:
    if current == target_head:
        return ()
    lower = current or "base"
    try:
        revisions = list(script.iterate_revisions(target_head, lower))
    except Exception:
        if current is None:
            return (target_head,)
        return (target_head,)
    # iterate_revisions walks down from upper; apply order is the reverse.
    ordered = [rev.revision for rev in reversed(revisions)]
    return tuple(ordered)


def apply_upgrade(
    *,
    forward_only: bool,
    alembic_config: Config | None = None,
    window: AppWindow | None = None,
    kinds: dict[str, str] | None = None,
) -> CompatDecision:
    cfg = _alembic_config(alembic_config)
    script = _script(cfg)
    target = window or load_window()
    kind_map = kinds or load_kinds()
    current = current_revision()
    pending = _pending_from_script(script, current, target.schema_head)
    decision = plan_upgrade(
        current_revision=current,
        window=target,
        kinds=kind_map,
        pending=pending,
        forward_only=forward_only,
        source_head=current,
    )
    if decision.action == "refuse":
        decision.outcome = "refused"
        return decision
    if decision.action == "noop":
        decision.outcome = "already_at_head"
        return decision
    command.upgrade(cfg, target.schema_head)
    decision.outcome = "applied"
    return decision


async def assert_servable() -> None:
    """Fail closed at process start when the live schema is older than min."""
    window = load_window()
    kinds = load_kinds()
    current = await current_revision_async()
    if not can_serve(current, window, kinds):
        raise RuntimeError(
            f"database schema {current!r} is below application min "
            f"{window.schema_min}; wait for the upgrade Job / curie-migrate "
            "phase, or restore a compatible backup"
        )


def wait_for_schema(*, attempts: int = 60, interval_s: float = 2.0) -> int:
    window = load_window()
    kinds = load_kinds()
    last: str | None = None
    for attempt in range(1, attempts + 1):
        try:
            last = asyncio.run(current_revision_async())
        except Exception as exc:
            last = None
            probe = type(exc).__name__
            if attempt == 1:
                print("Waiting for schema compatibility", file=sys.stderr)
            if attempt == attempts:
                print(
                    f"schema unavailable after {attempts} attempts; "
                    f"final probe error class: {probe}",
                    file=sys.stderr,
                )
                return 1
            time.sleep(interval_s)
            continue
        if can_serve(last, window, kinds):
            return 0
        if attempt == 1:
            print("Waiting for schema compatibility", file=sys.stderr)
        time.sleep(interval_s)
    print(
        f"schema {last!r} is below min {window.schema_min} after {attempts} attempts",
        file=sys.stderr,
    )
    return 1


def _forward_only_from_env() -> bool:
    raw = os.environ.get("CURIE_SCHEMA_FORWARD_ONLY", "")
    return raw.lower() in {"1", "true", "yes"}


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(prog="curie_api.schema_compat")
    sub = parser.add_subparsers(dest="cmd", required=True)
    sub.add_parser("plan", help="compare current database against this image")
    sub.add_parser("upgrade", help="plan then apply if allowed")
    sub.add_parser("wait", help="block until the live schema is servable")
    args = parser.parse_args(argv)
    forward_only = _forward_only_from_env()
    if args.cmd == "wait":
        return wait_for_schema()
    if args.cmd == "plan":
        cfg = _alembic_config()
        script = _script(cfg)
        window = load_window()
        current = current_revision()
        pending = _pending_from_script(script, current, window.schema_head)
        decision = plan_upgrade(
            current_revision=current,
            window=window,
            kinds=load_kinds(),
            pending=pending,
            forward_only=forward_only,
            source_head=current,
        )
        print(json.dumps(render_decision(decision), sort_keys=True))
        return 0 if decision.action != "refuse" else 2
    decision = apply_upgrade(forward_only=forward_only)
    print(json.dumps(render_decision(decision), sort_keys=True))
    if decision.action == "refuse":
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
